### ISSUE TITLE
Can't switch to GPU Eco Mode on G14 2023

### ISSUE THREAD

**Issue by Thrakios** [seerge/g-helper#345]
The bug is very simple.
Selecting Eco Mode does nothing other than freeze the UI for a few moments, then later return to Standard mode.

The log is attached.
[log.txt](https://github.com/seerge/g-helper/files/11432260/log.txt)

**Comment by seerge** [comment]
@Thrakios hello, i indeed see 
```
5/9/2023 11:32:06 PM: GPUEco = 1 : 0
```
in a log, app sends command and gets 0 as response (i.e. asus bios refuses to switch)

1. Can you try to run following in powershell (As admin), it should do same thing (disable GPU)
```
(Get-WmiObject -Namespace Root/WMI -Class AsusAtkWmi_WMNB).DEVS(0x00090020, 1)
```

2. What is your current nvidia optimus setting?

**Comment by seerge** [comment]
Oke, ``result : 0`` is pretty much same thing (bios refuses)

1. Can you try to set Optimus settings to actual "Optimus" (from auto) ? and see if it make a difference?
2. If doesn't help - try to set Optimus - to Nvidia GPU only, and try again?

**Comment by Thrakios** [comment]
1. There is no "Optimus" setting in my Nvidia Control Panel - Manage 3D settings options.
2. Nvidia GPU only doesn't fix the issue
3. There is no external screen connected

Oddly enough, it seem that after a reboot, G-Helper manages to disable the dGPU once, and only once.
You can see it at the very bottom of the log:
[log.txt](https://github.com/seerge/g-helper/files/11437655/log.txt)

**Comment by seerge** [comment]
@Thrakios 

Under optimus I meant this : 
![Screenshot 2023-05-10 110714](https://github.com/seerge/g-helper/assets/5920850/07e96af5-ec8f-4341-a7f8-eaa095d021eb)

From log i indeed see few times (i assume after restarts?) when eco command succeeds (in both directions even):
```
5/10/2023 11:57:47 AM: GPUEco = 1 : OK
...
5/10/2023 11:57:52 AM: GPUEco = 0 : OK
```

So the goal is to find what app (or driver) is blocking GPU from disabling after. 

1. What happens if you try Eco mode in armoury?
2. Can you click on GPU activity in system tray, and try to manually kill some of those processes? 
![Screenshot 2023-05-10 111132](https://github.com/seerge/g-helper/assets/5920850/19175ee1-3739-4b77-a531-e4c7c212b173)

**Comment by seerge** [comment]
@Thrakios Another thought, you can "reload" GPU driver in windows by pressing Ctr+Shift+WinKey+B . You screen will blink black for a second. Can you try that and disable gpu after?

**Comment by Thrakios** [comment]
I've tried Eco Mode both after enabling Optimus and Nvidia GPU only mode. It worked.
However from there, when I return to Standard Mode, then try to get back into Eco Mode it doesn't work properly:

- The Manage Display Mode disappears from NVIDIA Control Panel
- G-Helper gets stuck trying to get into Eco-Mode, and eventually returns to Standard Mode.
- There is no GPU activity report

![Screenshot 2023-05-10 181737](https://github.com/seerge/g-helper/assets/133026331/1f9b1053-e5af-43ef-acb6-171cc0db2d99)

**Comment by seerge** [comment]
@Thrakios  P.S. I have noticed you have AMD Adrenaline running (in tray), can you remove it from startup completely (from task manager)? May be amd is messing that ...

**Comment by Thrakios** [comment]
So : 
1 - Disabling just AMD Adrenaline did not fix the issue.
2 - Disabling then enabling the dGPU and then quickly enabling GPU Eco Mode works ! Finally there's something that works.

Yet Nvidia GPU activity (which started working) said there are no processes running, and still I couldn't switch without disabling and enabling it.

I'll try to uninstall-reinstall drivers now!

**Comment by seerge** [comment]
@Thrakios ok, that's cool :) let me know if driver reinstall helps

**Comment by Thrakios** [comment]
@seerge Unfortunately the driver reinstall didn't help :(

**Comment by Thrakios** [comment]
Thanks! Unfortunately it didn't work.
I tried twice, the second time running it with administrator privileges but it didn't fix it.
[log.txt](https://github.com/seerge/g-helper/files/11440796/log.txt)

**Comment by seerge** [comment]
@Thrakios and now : 
[GHelper.zip](https://github.com/seerge/g-helper/files/11440993/GHelper.zip)

**Comment by Thrakios** [comment]
@seerge 
Here's the log, didn't work unfortunately
[log.txt](https://github.com/seerge/g-helper/files/11441285/log.txt)

**Comment by seerge** [comment]
Oke

```
5/10/2023 8:24:43 PM: GPUEco = 1 : 0
5/10/2023 8:24:43 PM: Trying to disable GPU in a hard way
5/10/2023 8:24:43 PM: Device ID:PCI\VEN_10DE&DEV_27E0&SUBSYS_182D1043&REV_A1
5/10/2023 8:24:46 PM: Microsoft PnP Utility

Restarting device:         PCI\VEN_10DE&DEV_27E0&SUBSYS_182D1043&REV_A1\4&2b5a5da1&0&0009
Device restarted successfully.


5/10/2023 8:25:36 PM: GPUEco = 1 : 0
```

So it actually managed to restart device , but eco is still refused.

Can you try to run manually in powershell as admin

`` pnputil /restart-device /device-id "PCI\VEN_10DE&DEV_27E0&SUBSYS_182D1043&REV_A1\4&2b5a5da1&0&0009" ``

And then try to set Eco mode in app again ?

**Comment by seerge** [comment]
It could be that it needs to be disabled -> enabled (as you did manually), instead of restarting

For that you can try : [GHelper.zip](https://github.com/seerge/g-helper/files/11441334/GHelper.zip)

**Comment by Thrakios** [comment]
I manually tried the powershell command, unfortunately it didn't work.
I also tried the new release, here's the log. Still did not work :/

[log.txt](https://github.com/seerge/g-helper/files/11441367/log.txt)

**Comment by seerge** [comment]
@Thrakios does manually disabling / enabling GPU in device manager (visually, as you originally did) and then setting Eco still work for you ? Or it worked only once ? :|

**Comment by seerge** [comment]
@Thrakios and if manual disabling doesn't work anymore, I thought about 2 other general things to chech/try

1. Rest that I suggested to [Avacado_I_Guess](https://www.reddit.com/user/Avacado_I_Guess/) from reddit : You may want to do a reset : turn off laptop normally, press and hold power button for like 30-40 seconds, then boot it again (it will take a bit slower to boot). This will reset all hardware settings

2. Check if fast boot is ON or OFF in bios settings (F2 on boot). And try to set opposite (i.e. if it's ON turn it OFF)

**Comment by Thrakios** [comment]
Manually disabling / enabling GPU does still allow to enable Eco mode.
I guess both disabling, and enabling takes a bit of time. Trying to time it, I had 3:30 seconds to disable, and 2:15 seconds to enable. Then I set Eco Mode almost immediately (within 2 seconds).

**Comment by seerge** [comment]
@Thrakios oke, try version with 2 seonds delays 
[GHelper.zip](https://github.com/seerge/g-helper/files/11441740/GHelper.zip)

**Comment by Thrakios** [comment]
Still did not work unfortunately.
It did work twice in a row at first so I was quite excited, but it seems it was a fluke. When I tried again it didn't work.
[log.txt](https://github.com/seerge/g-helper/files/11441826/log.txt)

**Comment by seerge** [comment]
You can try to find appropriate delays by running 3 commands (with delay) in powershell as admin :

```
pnputil /disable-device /device-id "PCI\VEN_10DE&DEV_27E0&SUBSYS_182D1043&REV_A1\4&2b5a5da1&0&0009"
pnputil /enable-device /device-id "PCI\VEN_10DE&DEV_27E0&SUBSYS_182D1043&REV_A1\4&2b5a5da1&0&0009"
(Get-WmiObject -Namespace Root/WMI -Class AsusAtkWmi_WMNB).DEVS(0x00090020, 1)
```

**Comment by seerge** [comment]
@Thrakios or try build with 4 second delays between each : 
[GHelper.zip](https://github.com/seerge/g-helper/files/11441967/GHelper.zip)

**Comment by Thrakios** [comment]
The commands, when I input them manually to powershell in Admin Mode don't seem to disable the dGPU. 😅

Here's a screen recording where I do the steps manually, hoping it may help. 

https://github.com/seerge/g-helper/assets/133026331/dabb5b4f-7816-4f96-9a21-86ab24c00064

**Comment by seerge** [comment]
@Thrakios then i can only blame his majesty windows, as pnputil is supposed (and actually seem) to do exactly same as you do manually in device manager.

Do you run powershell commands with delay? Try to run them with big delays, like 10 seconds or so. If it doesn't work, i can't help here . At least there is some workaround.

Did you had a chance to try bios reset / fast boot things i mentioned before ?

**Comment by Thrakios** [comment]
PS : I did try with disabling fast boot, and had done a reset yesterday

**Comment by seerge** [comment]
@Thrakios or just try this build (i have extended timings delays even more) : 
[GHelper.zip](https://github.com/seerge/g-helper/files/11444273/GHelper.zip)

**Comment by pettijohn** [comment]
I've experienced this same issue from original post on a 2023 G14 with RTX 4070. I wrote this powershell script:

```
$device = Get-PnpDevice | Where-Object { $_.FriendlyName -imatch 'NVIDIA' -and $_.Class -eq 'Display' }
Disable-PnpDevice $device.InstanceId -Confirm:$true
Enable-PnpDevice $device.InstanceId -Confirm:$true
```

I assume the drivers are buggy and don't seem to properly "release" when programs are done, and g-helper hangs as discussed. My workaround is to leave the device on Eco mode except when I'm actively using the GPU. When I'm done using the GPU, I quit all relevant programs, run my script, and then enable Eco mode in G-helper. It's far from ideal but a suitable workaround while we wait for better drivers from bleeding edge hardware. 

Related, I have a bug where the device bugchecks (BSOD) when trying to sleep while an external display is connected to the NVIDIA GPU. I was able to capture a memory dump and NVIDIA confirmed they can reproduce the issue, so it's a matter of time until they triage & fix that issue.

**Comment by Thrakios** [comment]
@pettijohn Your powershell script worked! Whereas I didn't manage to get the other one disabling my GPU for some reason.

**Comment by seerge** [comment]
@Thrakios @pettijohn dis you try build from my last post, and if so cam you post logs from it after trying eco?

**Comment by seerge** [comment]
Just in case I have pushed 0.60 https://github.com/seerge/g-helper/releases/tag/v0.60 (with pnputil enable / disable) to collect more feedback.

**Comment by pettijohn** [comment]
Tried to switch to Eco mode at about 2:15. Tried twice in the app, never succeeded. Then I ran my above powershell script, started g-helper, and successfully switched to Eco mode. I also included a few exceptions from the logs. 
[v060.log](https://github.com/seerge/g-helper/files/11456654/v060.log)

**Comment by pettijohn** [comment]
Tried to switch to Eco mode at about 2:15. Tried twice in the app, never succeeded. Then I ran my above powershell script, started g-helper, and successfully switched to Eco mode. I also included a few exceptions from the logs.

[v060-and-exceptions.log](https://github.com/seerge/g-helper/files/11456660/v060-and-exceptions.log)

**Comment by seerge** [comment]
@pettijohn i see that nvidia API is completely turned off in your case. Looks like driver in windows is acting in aftermath from switching eco / standard ?

**Comment by Neristee** [comment]
Don't want to clutter the issue report too much. Just providing some more infos (maybe).

This whole issue is awfully reminding me of the troubles I currently have with eco mode (6900HS, 3070Ti), even with armoury crates installed. My feedback is kinda out of scope since I'm using an Insider version of Windows. But it might maybe help ?

I tried numerous tricks to enable it, including disabling manually the nvidia dGPU... and when I successfully trick windows enough to be able to turn eco mode on.. I get a 100% chance of a BSOD mentioning : POWER_STATE_FAILURE from the nvidia driver. 

Sometimes rebooting from this BSOD will have eco mode activated, but the dGPU would be invisible in device manager, no nvidia control panel... and re-bsod... until I turn on standard mode (if I have the time during before the BSOD).

This might very well be a Windows/Nvidia issue (i checked old nvidia drivers, didn't change much... so probably a windows issue).

**Comment by pettijohn** [comment]
@Neristee if you can 100% create a BSOD, submit the memory dump to NVIDIA, see here: https://nvidia.custhelp.com/app/answers/detail/a_id/4842

**Comment by seerge** [comment]
@Neristee changes from that build are live in 0.61 already
@pettijohn / @Thrakios did you had chance to try it ? Does it also work for you same as it does for @Neristee ? I will close this issue then :)

**Comment by Neristee** [comment]
### Concerns this issue

Still using ECO mode through GHelper. No BSOD. And battery discharge numbers I haven't seen in a long time.  

![Screenshot 2023-05-13 192339](https://github.com/seerge/g-helper/assets/15640409/867d0fcc-e0a6-4811-b1d6-f909a64ea348)


### Doesn't really concerns this issue

> @Neristee if you can 100% create a BSOD, submit the memory dump to NVIDIA, see here: https://nvidia.custhelp.com/app/answers/detail/a_id/4842

I can't reproduce the bsod on the current windows insider canary build. I was able to reproduce it on last week's build though. I didn't even try tbh because I just got used to use "Standard" at the expense of battery life. To be fair, there seems to have a lot of USB related work under the hood (xbox controller wasn't working last build for example) so it might be a me issue on that one.

**Comment by Thrakios** [comment]
Yes the dGPU area in my AC looks the same. Though I've now uninstalled it.
As for GHelper, here's a log. I didn't manage to get it working unfortunately :(

[log.txt](https://github.com/seerge/g-helper/files/11470968/log.txt)

**Comment by seerge** [comment]
@Thrakios that's sad, but i see that it seem to restart gpu (and uses powershell way now with 2 seconds pause inbetween disabling and enabling) :) then i don't know what to do in your case

**Comment by seerge** [comment]
Btw, i hope no one here uses asus smart display service , as this issue has reminded me that in certain cituations it can also interfere https://github.com/seerge/g-helper/issues/377

**Comment by Thrakios** [comment]
@seerge I have uninstalled ASUS Smart Display Control, and now GHelper does successfully get into Eco Mode, after the automatic GPU restart.

**Comment by seerge** [comment]
@Thrakios hooray!

**Comment by Knox623** [comment]
@seerge Hi, I unfortunately seem to be having the same issue originally stated. I've double checked and I do not have ASUS Smart Display Control installed.
[config.txt](https://github.com/seerge/g-helper/files/11537779/config.txt)
[log.txt](https://github.com/seerge/g-helper/files/11537780/log.txt)

**Comment by seerge** [comment]
@Knox623 log shows responses from your BIOS :) it says that gpu disable command was successfull + then when app double checked - it still reported that it's off. So programmatically your computer definetely thinks that GPU is off on this stage.

Temperature has nothing to do with it - as it's asus sensor on motherboard, it reports temperature even if GPU is powered off

How does the "task manager" connected to GPU ? Does GPU show up in "device manager"?

**Comment by Knox623** [comment]
@seerge Can I ask how quickly after you double check the GPU has been disabled? Because for a brief amount of time the GPU is disabled (this is when it restarts itself in the device manager) though shortly after it reenables itself. It takes maybe 10 seconds to show up again. 

The GPU shows up in both device manager and task manager.

**Comment by Knox623** [comment]
> @seerge Can I ask how quickly after you double check the GPU has been disabled? Because for a brief amount of time the GPU is disabled (this is when it restarts itself in the device manager) though shortly after it reenables itself. It takes maybe 10 seconds to show up again.
> 
> The GPU shows up in both device manager and task manager.

![image](https://github.com/seerge/g-helper/assets/63962644/760a4f5b-6cdd-457e-81d6-69a25036d12f)

**Comment by Knox623** [comment]
> Temperature has nothing to do with it - as it's asus sensor on motherboard, it reports temperature even if GPU is powered off

Also as for this bit, when the GPU is successfully disabled GHelper doesn't seem to display a temperature. This is my gauge for whether its on or not.
![image](https://github.com/seerge/g-helper/assets/63962644/3cb1afd7-b0f9-445a-88b9-e7ecde047614)

**Comment by Knox623** [comment]
@seerge Yeah sorry forgot to mention that. No worries I suppose I'll just switch back to armoury crate and see if that can disable the gpu properly.

**Comment by seerge** [comment]
To everyone involved @Knox623 @Thrakios @pettijohn @Neristee 

I have prepared a new build that can kill all processes using GPU before trying to set Eco. You need to go Extra -> select "Stop all apps using GPU before switching to Eco" for that. I'm wondering if it will "help" in your situations 

[GHelper.zip](https://github.com/seerge/g-helper/files/11591069/GHelper.zip)

Let me know (and post logs if possible). Thanks.

**Comment by Jokerr2** [comment]
It works (kinda), well after trying to switch to eco mode in g helper, it does restart the GPU but after the GPU restarts it does not keep the GPU disabled, the GPU gets enabled again after the restart. The way I have got it to stay disabled is  by switching to eco mode and then immediately trying to switch back to standard mode after GPU starts to restart and back to eco mode while the GPU is in the process of restarting which keeps the GPU in eco mode after it restarts. It does not have a 100% success rate, probably more like 60 or 70 %, but at least something works

**Comment by seerge** [comment]
@Jokerr2 I **wouldn't advice** all this strange operations (with disabling then immediate enabling) honestly. 

From reddit I see that latest bios update (your case too?) seem to introduce new problems for eco mode https://www.reddit.com/r/ZephyrusG14/comments/142iipu/eco_mode_not_working_after_bios_309_update/  People in that topic use Armoury, but since problem is with bios - it could affect everyone on that version.

**Comment by Zandexstudio** [comment]
Hi, Can someone help me with my laptop. My 3060 Graphics card stopped working all of a sudden. It is an asus G15. I did a clean installation of drivers, tried different drivers, also did a clean windows installation. Nothing seems to help me, can any of you help me please? it say error 43. For some reason my integrated won't switch to dedicated. Bios shows 3060. But cpuz and other softwares including nvidea control panel don't recognise my 3060 graphics. I tried switching to standard gpu mode using ghelper and also armoury crate but it never switch. It shows switching like forever

**Comment by seerge** [comment]
@Zandexstudio you can try hardware reset https://github.com/seerge/g-helper#how-do-i-do-a-hardware-reset-on-a-laptop

**Comment by joe-700** [comment]
I have a similar issue switching to eco mode on 2023 G14 (4080 version). I have had the problem with armoury crate before moving to the G-Helper and its taken a while to get the issue to reliably repeat.
Issue:
I noticed that occasionally Armoury Crate would not switch to eco mode and the 'Manage display mode' options in Nvidia control panel disappeared. stopping GPU applications would not help and several restarts or cycling the Ultimate mode seemed to fix the problem, however it was difficult to get the problem to reliably repeat. I then uninstalled Armory Crate and any AMD software, fired up G-Helper and set it to stop all Asus services and stop all apps using GPU when switching to eco mode. I now have 0 Asus services running and the issue is now repeatable. This is what happens:
1. Restart the PC from eco mode
2. Noticed that the Optimus and Nvidia applications are not in the system tray on restart (this is different to a restart from 'standard' mode)
3. Toggle to standard mode, now Nvidia and Optimus system tray apps start-up
4. Open Nvidia control panel and notice that the MDM option is missing
5. I then have to toggle back to eco mode, where the system hangs up for a while before giving me the option to restart DGPU in device manager.
6. Click yes to restart DGPU
7. now G-Helper will switch into Eco mode and moving back into standard mode brings back the MDM in the Nvidia control panel.
8. All is well, happily switching back and forth all day without issue until the system is restarted from eco mode again, then you have to repeat steps 1-7 again to fix.

I have the latest BIOS 310

This only happens when you start the system into the eco mode state. I'm guessing that the DGPU is not starting correctly when switching to standard mode. 

This is a fantastic program, way better that Armoury Crate, I will continue to use this going forward but I would love to be able to start the computer from eco mode without going through the above routine to get everything running correctly.

**Comment by seerge** [comment]
@joe-700 hello, well, as mentioned in this topic. Enabling/disabling GPU is literally 1 simple command to bios (and the only existing) 

0 - to enable gpu, 1 to disable
``
(Get-WmiObject -Namespace Root/WMI -Class AsusAtkWmi_WMNB).DEVS(0x00090020, 1)
``

App can do only that :) and does. 

If it doesn't work sometimes (from your words also in Armoury), it looks like there are still issues with BIOS and I hope they will eventually fix it. 

I can only suggest to try to do a hardware reset as I was mentioning before, as it re-enables dGPU as well
https://github.com/seerge/g-helper#how-do-i-do-a-hardware-reset-on-a-laptop

**Comment by joe-700** [comment]
I suspect your right, fingers crossed it gets addressed in future BIOS updates. 

Fantastic little tool BTW, does everything AC does without the nonsense.

**Comment by Zandexstudio** [comment]
I should try this. This issue happened ri8 after the expiration of
warranty. Btw was the nvidea driver showing error code 43 in your case?
